tensorflow/contrib/cluster_resolver/python/training/slurm_cluster_resolver.py

```diff
@@ -0,0 +1,180 @@
+
+    Args:
+      jobs: Dictionary with job names as key and number of tasks in the job as
+        value
+      port_base: The first port number to start with for processes on a node.
+      gpus_per_node: Number of GPUs avaliable on each node.
+      gpus_per_task: Number of GPUs to be used for each task.
+      tasks_per_node: Number of tasks to run on each node, if not set defaults
+        to Slurm's output environment variable SLURM_NTASKS_PER_NODE.
+      auto_set_gpu: Set the visible CUDA devices automatically while resolving
+        the cluster by setting CUDA_VISIBLE_DEVICE environment variable.
+        Defaults to True.
+
+    Raises:
+      RuntimeError: If requested more GPUs per node then avaliable or requested
+      more tasks then assigned tasks.
+    """
+
+    # check if launched by mpirun
+    if 'OMPI_COMM_WORLD_RANK' in os.environ:
+      self._rank = int(os.environ['OMPI_COMM_WORLD_RANK'])
+      num_tasks = int(os.environ['OMPI_COMM_WORLD_SIZE'])
+    else:
+      self._rank = int(os.environ['SLURM_PROCID'])
+      num_tasks = int(os.environ['SLURM_NTASKS'])
+
+    self._jobs = jobs
+    self._port_base = port_base
+
+    # user specification overrides SLURM specification
+    if tasks_per_node is not None:
+      self._tasks_per_node = tasks_per_node
+    elif tasks_per_node is None and 'SLURM_NTASKS_PER_NODE' in os.environ:
+      self._tasks_per_node = int(os.environ['SLURM_NTASKS_PER_NODE'])
+    else:
+      raise RuntimeError('Neither `tasks_per_node` or \
+                          SLURM_NTASKS_PER_NODE is set')
+
+    self._gpus_per_node = gpus_per_node
+    self._gpus_per_task = gpus_per_task
+
+    self._auto_set_gpu = auto_set_gpu
+    self._job_name = None
+    self._task_index = None
+
+    self._gpu_allocation = []
+    self._cluster_allocation = {}
+
+    if self._tasks_per_node * self._gpus_per_task > self._gpus_per_node:
+      raise RuntimeError('Requested more GPUs per node then avaliable')
+
+    if sum(self._jobs.values()) != num_tasks:
+      raise RuntimeError('Requested more tasks then assigned tasks')
+
+  def cluster_spec(self):
+    """Returns a ClusterSpec object based on the latest instance group info.
+
+    This returns a ClusterSpec object for use based on information from the
+    specified initialization parameters and Slurm environment variables. The
+    cluster specification is resolved each time this function is called. The
+    resolver extract hostnames of nodes by scontrol and pack tasks in that
+    order until a node a has number of tasks that is equal to specification.
+    GPUs on nodes are allocated to tasks by specification through setting
+    CUDA_VISIBLE_DEVICE environment variable.
+
+    Returns:
+      A ClusterSpec containing host information retrieved from Slurm's
+        environment variables.
+    """
+    hostlist = subprocess.check_output(['scontrol', 'show', 'hostname']).\
+               decode("utf-8").strip().split('\n')
```

Comment:
Be consistent in quote usage (either all `"` or all `'`)

Reply:
I wonder if you want to wrap the call to scontrol in a private method? Also is there another way to determine this other than running a subprocess (e.g. a more documented API)

This will also make your test below easier rather than patching out the subprocess.

Reply:
Created a private method that directly return a list of node names and modified the unit tests accordingly to patch that function instead of patching the whole subprocess.

In fact, another way of extracting the node names will be to read an environment variable called ```SLURM_NODELIST```. However the draw back is that the list returned will be in compressed format. i.e. the lsit of nodes ```t02n13,t02n14,t02n15,t03n02,t03n03``` will be returned as ```t02n[13-15],t03n[02-03]```, meaning that parsers will have to be written. A python package https://www.nsc.liu.se/~kent/python-hostlist/ can be used to expand these lists but it introduces additional dependency and the functions provided maybe overkill for this, but it can definitely be changed in the future!